File: spring-data-requery/src/test/java/org/springframework/data/requery/mapping/RequeryPersistentPropertyTest.java
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.mapping;

import io.requery.Embedded;
import io.requery.OneToOne;
import io.requery.Transient;
import io.requery.meta.EntityModel;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.springframework.data.annotation.Version;
import org.springframework.data.util.ClassTypeInformation;
import org.springframework.data.util.TypeInformation;

import javax.persistence.Access;
import javax.persistence.AccessType;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * org.springframework.data.requery.mapping.RequeryPersistentPropertyTest
 *
 * @author debop
 * @since 18. 6. 8
 */
@RunWith(MockitoJUnitRunner.class)
public class RequeryPersistentPropertyTest {

    @Mock EntityModel entityModel;

    RequeryMappingContext context;
    RequeryPersistentEntity<?> entity;

    @Before
    public void setup() {
        context = new RequeryMappingContext();
        entity = context.getRequiredPersistentEntity(Sample.class);
    }

    @Test
    public void considersOneToOneMappedPropertyAssociation() {
        RequeryPersistentProperty property = entity.getRequiredPersistentProperty("other");
        assertThat(property).isNotNull();
        assertThat(property.isAssociation()).isTrue();
    }

    @Test
    public void considersTransientFieldsAsTransient() {
        RequeryPersistentProperty property = entity.getPersistentProperty("transientProp");
        assertThat(property).isNull();
    }

    @Test
    public void considersEmbeddedablePropertyAnAssociation() {
        assertThat(context.getPersistentEntity(SampleEmbeddable.class)).isNotNull();
    }

    @Test
    public void doesNotConsiderAnEmbeddablePropertyAnAssociation() {
        assertThat(entity.getPersistentProperty("embeddable").isAssociation()).isFalse();
    }

    @Test
    public void doesNotConsiderAnEmbeddedPropertyAnAssociation() {
        assertThat(entity.getPersistentProperty("embedded").isAssociation()).isFalse();
    }

    @Test
    public void considersPropertyLevelAccessTypeDefinitions() {
        assertThat(getProperty(PropertyLevelPropertyAccess.class, "field").usePropertyAccess()).isFalse();
        assertThat(getProperty(PropertyLevelPropertyAccess.class, "property").usePropertyAccess()).isTrue();
    }

    @Test
    public void propertyLevelAccessTypeTrumpsTypeLevelDefinition() {
        assertThat(getProperty(PropertyLevelDefinitionTrumpsTypeLevelOne.class, "field").usePropertyAccess()).isFalse();
        assertThat(getProperty(PropertyLevelDefinitionTrumpsTypeLevelOne.class, "property").usePropertyAccess()).isTrue();

        // JPA 는 상관하지 않습니다^^
        assertThat(getProperty(PropertyLevelDefinitionTrumpsTypeLevelOne2.class, "field").usePropertyAccess()).isTrue();
        assertThat(getProperty(PropertyLevelDefinitionTrumpsTypeLevelOne2.class, "property").usePropertyAccess()).isTrue();
    }

    @Test
    public void considersJpaAccessDefinitionAnnotations() {
        assertThat(getProperty(TypeLevelPropertyAccess.class, "id").usePropertyAccess()).isTrue();
    }

    @Test // DATAJPA-619
    public void springDataAnnotationTrumpsJpaIfBothOnTypeLevel() {
        assertThat(getProperty(CompetingTypeLevelAnnotations.class, "id").usePropertyAccess()).isFalse();
    }

    @Test // DATAJPA-619
    public void springDataAnnotationTrumpsJpaIfBothOnPropertyLevel() {
        assertThat(getProperty(CompetingPropertyLevelAnnotations.class, "id").usePropertyAccess()).isFalse();
    }

    @Test
    public void detectRequeryVersionAnnotation() {
        assertThat(getProperty(RequeryVersioned.class, "id").isIdProperty()).isTrue();

        assertThat(getProperty(SpringDataVersioned.class, "version").isVersionProperty()).isFalse();
        assertThat(getProperty(RequeryVersioned.class, "version").isVersionProperty()).isTrue();
    }

    @Test
    public void considersTargetEntityTypeForPropertyType() {
        RequeryPersistentProperty property = getProperty(SpecializedAssociation.class, "api");

        assertThat(property.getType()).isAssignableFrom(Api.class);

        Iterable<? extends TypeInformation<?>> entityTypes = property.getPersistentEntityTypes();
        assertThat(entityTypes.iterator().hasNext()).isTrue();
        assertThat(entityTypes.iterator().next()).isEqualTo(ClassTypeInformation.from(Api.class));
    }


    private RequeryPersistentProperty getProperty(Class<?> ownerType, String propertyName) {
        RequeryPersistentEntity<?> entity = context.getRequiredPersistentEntity(ownerType);
        return entity.getRequiredPersistentProperty(propertyName);
    }

    static class Sample {

        @OneToOne Sample other;
        @Transient String transientProp;

        SampleEmbeddable embeddable;

        @Embedded
        SampleEmbedded embedded;
    }

    @Embedded
    static class SampleEmbeddable {
    }

    static class SampleEmbedded {
    }

    @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.PROPERTY)
    static class TypeLevelPropertyAccess {
        private String id;

        public String getId() { return id; }
    }

    static class PropertyLevelPropertyAccess {

        String field;
        String property;

        @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.PROPERTY)
        public String getProperty() {
            return property;
        }
    }

    @Access(AccessType.FIELD)
    static class PropertyLevelDefinitionTrumpsTypeLevelOne {
        String field;
        String property;

        @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.PROPERTY)
        public String getProperty() {
            return property;
        }
    }

    @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.PROPERTY)
    static class PropertyLevelDefinitionTrumpsTypeLevelOne2 {
        @Access(AccessType.FIELD)
        String field;

        String property;

        public String getProperty() {
            return property;
        }
    }

    @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.FIELD)
    @Access(AccessType.PROPERTY)
    static class CompetingTypeLevelAnnotations {

        private String id;

        public String getId() { return id; }
    }

    @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.FIELD)
    @Access(AccessType.PROPERTY)
    static class CompetingPropertyLevelAnnotations {

        private String id;

        @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.FIELD)
        @Access(AccessType.PROPERTY)
        public String getId() {
            return id;
        }
    }

    static class SpringDataVersioned {

        @Version long version;
    }

    static class RequeryVersioned {

        @io.requery.Key Long id;

        @io.requery.Version long version;
    }

    static class SpecializedAssociation {

        @io.requery.ManyToOne Api api;
    }

    static interface Api {}

    static class Implementation {}

    static class WithIndexed {
        @io.requery.Column(index = true) String name;
        String updatable;
    }
}

File: spring-data-requery-kotlin/src/test/kotlin/org/springframework/data/requery/kotlin/mapping/RequeryPersistentPropertyTest.kt
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.kotlin.mapping

import com.nhaarman.mockito_kotlin.mock
import io.requery.Column
import io.requery.Embedded
import io.requery.Key
import io.requery.ManyToOne
import io.requery.OneToOne
import io.requery.Transient
import io.requery.meta.EntityModel
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import org.springframework.data.annotation.Version
import org.springframework.data.util.ClassTypeInformation

/**
 * RequeryPersistentPropertyTest
 *
 * @author debop
 */
@RunWith(MockitoJUnitRunner::class)
class RequeryPersistentPropertyTest {

    val entityModel = mock<EntityModel>()

    lateinit var context: RequeryMappingContext
    lateinit var entity: RequeryPersistentEntity<*>

    @Before
    fun setup() {
        context = RequeryMappingContext()
        entity = context.getRequiredPersistentEntity(Sample::class.java)
    }

    @Test
    fun `considers OneToOne mapped property association`() {
        val property = entity.getRequiredPersistentProperty("other")
        assertThat(property).isNotNull
        assertThat(property.isAssociation).isTrue()
    }

    @Test
    fun `considers transient fields as transient`() {
        val property = entity.getPersistentProperty("transientProp")
        assertThat(property).isNull()
    }

    @Test
    fun `considers embeddable property an assosication`() {
        assertThat(context.getPersistentEntity(SampleEmbeddable::class.java)).isNotNull
    }

    @Test
    fun `does not consider embeddable property assocation`() {
        assertThat(entity.getPersistentProperty("embeddable")!!.isAssociation).isFalse()
    }

    @Test
    fun `does not consider embeddable property an assocation`() {
        assertThat(entity.getPersistentProperty("embedded")!!.isAssociation).isFalse()
    }

    @Test
    fun `detect Requery Version Annotation`() {
        assertThat(getProperty(RequeryVersioned::class.java, "id").isIdProperty).isTrue()

        assertThat(getProperty(SpringDataVersioned::class.java, "version").isVersionProperty).isFalse()
        assertThat(getProperty(RequeryVersioned::class.java, "version").isVersionProperty).isTrue()
    }

    @Test
    fun `considers target entity type for property type`() {

        val property = getProperty(SpecializedAssociation::class.java, "api")

        val entityTypes = property.persistentEntityTypes
        assertThat(entityTypes.iterator().hasNext()).isTrue()
        assertThat(entityTypes.iterator().next()).isEqualTo(ClassTypeInformation.from(Api::class.java))
    }


    private fun getProperty(ownerType: Class<*>, propertyName: String): RequeryPersistentProperty {
        val entity = context.getRequiredPersistentEntity(ownerType)
        return entity.getRequiredPersistentProperty(propertyName)
    }

    class Sample {

        @get:OneToOne
        var other: Sample? = null

        @get:Transient
        var transientProp: String? = null

        var embeddable: SampleEmbeddable? = null

        @get:Embedded
        var embedded: SampleEmbedded? = null
    }

    @Embedded
    class SampleEmbeddable

    class SampleEmbedded

    @org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.PROPERTY)
    class TypeLevelPropertyAccess {

        var id: String? = null
    }

    class PropertyLevelPropertyAccess {

        var field: String? = null
        @get:org.springframework.data.annotation.AccessType(org.springframework.data.annotation.AccessType.Type.PROPERTY)
        var property: String? = null
    }


    class SpringDataVersioned {

        @Version var version: Long = 0
    }

    class RequeryVersioned {

        @get:Key
        var id: Long? = null

        @io.requery.Version
        var version: Long = 0
    }

    class SpecializedAssociation {

        @get:ManyToOne
        var api: Api? = null
    }

    interface Api

    class Implementation

    class WithIndexed {

        @get:Column(index = true)
        var name: String? = null

        var updatable: String? = null
    }
}